### DISCUSSION TITLE
exploring-dotnet-6-part-6-supporting-integration-tests-with-webapplicationfactory-in-dotnet-6/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#44] [Announcements]
# exploring-dotnet-6-part-6-supporting-integration-tests-with-webapplicationfactory-in-dotnet-6/

In this post I look at the changes that were made to WebApplicationFactory<T> to support minimal hosting APIs with WebApplicationBuilder in .NET 6

https://andrewlock.net/exploring-dotnet-6-part-6-supporting-integration-tests-with-webapplicationfactory-in-dotnet-6/

**Comment by alex-ht-work** [comment]
What I like a lot about the generic host model is that it minimizes the amount of "hidden magic" behind the application startup process: all that happens are explicit method calls on explicitly present objects. That means, in my tests I can do the same: create a host builder, configure it, build the host, start the host. I can re-use the code from my `Program.Main` as much as I need. For example, I can have a `Program.BuildHost` method, called both from `Program.Main` and from the test, but customize it with parameters, because the test host is not exactly the same as the program host. First of all, I do not want the test host to pick up config files from the main program, but rather supply a separate test-specific config. Then I may want to replace the logging framework, redirecting logs to test output. I can also skip components not related to the test... There may be other modifications needed. It all works with a generic host. Brilliant.

Sadly, things are not that smooth with web applications. The main reason, IMO, is that ASP.NET developers aim to make the framework "easy to start", introducing back that "hidden magic".

I tried the same approach (that work for me with the generic host) to a .NET 6 style web API, and it almost worked!.. Almost. Except my controllers could not be discovered. So if I use `app.MapGet()` it works, but as soon as I switch to controllers, I get 404 error. Need to note, that I was not using `WebApplicationFactory` nor `TestServer`. I was making actual HTTP calls to my application on `localhost`. I guess the controller discovery magic does not work when I try to engage it from a test assembly... That is how I came to use `WebApplicationFactory`, which looks promising because it is there for exactly this purpose: run a test host.

The problem is, I can not figure out how to achieve the same degree of customization as I had before. There seem to be ways to customize `WebApplicationFactory`, like override `IHost CreateHost(IHostBuilder builder)` method... This allows me to completely replace the host configuration or (calling a base class implementation) get the same I have in the main program. But what I need is to make several modifications to the main program host configuration. Can I?... Looks like I can add something to the builder, but how to "not add" things I do not want? What is exactly the way customization works? None of the documentation answers that, and trying to trace into `WebApplicationFactory`'s guts, I face all the complexity of its implementation, which is ugh... impressive.

I must say, this article is awesome, it is the best explanation I could find, but it still does not answer the question how to customize the `WebApplicationFactory` properly. This would make another very useful article, if you can do it...

So far, the only workaround that comes to my mind is to set some parameters for my host building function as static variables in the `Program` class. That way I can set them from my test and then use `WebApplicationFactory` as it is. Not the most elegant approach...

Thank you again for the article.

**Reply by BrunoJuchli** [reply]
@alex-ht-work
I was faced with the same problem. Turns out, it's difficult to read and understand the source of `WebApplicationFactory`, but the code to get it running for .Net 6/7 web app is actually not that difficult:

- extract stuff from program.cs which is supposed to be reused in unit tests into methods, which in my case is everything apart from configuration
- create an xUnit fixture as a Replacement for `WebApplicationFactory` (i'm using xUnit).
  - i created a base class which contains the logic which replaces `WebApplicationFactory`.
    Here I chose to copy a bit of code from the `WebApplicationFactory` repository.
  - and a derived class which contains the specifics for the System-under-test

This is how it looks:

### Program cs

    var builder = WebApplication.CreateBuilder(args);

    var config = /* reading the config */;

    ConfigureApplication(
        builder,
        config);

    var app = builder.Build();

    ConfigurePipeline(app);

    app.Run();

## xUnit Fixture

### Base Class

    using Microsoft.AspNetCore.Builder;
    using Microsoft.AspNetCore.Hosting;
    using Microsoft.AspNetCore.Hosting.Server;
    using Microsoft.AspNetCore.Mvc.Testing.Handlers;
    using Microsoft.AspNetCore.TestHost;
    using Microsoft.Extensions.DependencyInjection;
    using Microsoft.Extensions.Hosting;

    namespace WebApiServer.SystemTests;

    public abstract class AbstractTestServerFixture :
        IAsyncLifetime
    {
        private readonly CancellationTokenSource cancellationTokenSource;
        private readonly List<HttpClient> clients;
        private WebApplication? app;
        private TestServer? server;

        protected AbstractTestServerFixture()
        {
            cancellationTokenSource = new();
            clients = new();
        }

        async Task IAsyncLifetime.InitializeAsync()
        {
            var builder = WebApplication.CreateBuilder(
                new WebApplicationOptions
                {
                    EnvironmentName = Environments.Development
                });
            builder.WebHost
                .UseShutdownTimeout(TimeSpan.FromSeconds(5))
                .UseTestServer();
            
            await ConfigureApplication(builder);

            app = builder.Build();
            
            await ConfigurePipeline(app);

            server = (TestServer)app.Services.GetRequiredService<IServer>();
            
            await app.StartAsync(cancellationTokenSource.Token);
        }

        protected abstract Task ConfigureApplication(WebApplicationBuilder builder);

        protected abstract Task ConfigurePipeline(WebApplication app);

        async Task IAsyncLifetime.DisposeAsync()
        {
            foreach (var client in clients)
            {
                client.Dispose();
            }
            
            cancellationTokenSource.Cancel();

            app?.StopAsync(TimeSpan.FromSeconds(30));
            await (app?.DisposeAsync() ?? ValueTask.CompletedTask);
            server?.Dispose();
        }

        public HttpClient CreateClient()
        {
            return CreateClientInternal(
                new RedirectHandler(7),
                new CookieContainerHandler());
        }


        // MIT License
        // loosely based on the source code of https://www.nuget.org/packages/Microsoft.AspNetCore.Mvc.Testing
        // found at https://github.com/dotnet/aspnetcore/tree/main/src/Mvc/Mvc.Testing/src
        // (version at the time of writing: https://github.com/dotnet/aspnetcore/tree/cbb4916ecc63785267a750a78b42f3a769230509/src/Mvc/Mvc.Testing/src)
        private HttpClient CreateClientInternal(
            params DelegatingHandler[] handlers)
        {
            HttpClient client;
            if (handlers.Length == 0)
            {
                client = server!.CreateClient();
            }
            else
            {
                for (var i = handlers.Length - 1; i > 0; i--)
                {
                    handlers[i - 1].InnerHandler = handlers[i];
                }

                var serverHandler = server!.CreateHandler();
                handlers[^1].InnerHandler = serverHandler;

                client = new HttpClient(handlers[0]);
            }

            clients.Add(client);

            client.BaseAddress = server.BaseAddress;

            return client;
        }
    }

### Concrete Fixture Class

    using Microsoft.AspNetCore.Builder;
    using Microsoft.Extensions.Configuration;
    using Microsoft.Extensions.Hosting;
    using Xunit.Abstractions;

    namespace WebApiServer.SystemTests;

    public class WebApiServerFixture :
        AbstractTestServerFixture
    {
        protected override Task ConfigureApplication(
            WebApplicationBuilder builder)
        {
            var config = /* your test config */;
        
            Program.ConfigureApplication(
                builder,
                config);
            
            return Task.CompletedTask;
        }

        protected override Task ConfigurePipeline(
            WebApplication app)
        {
            Program.ConfigurePipeline(app);

            return Task.CompletedTask;
        }
    }


## Test

    namespace WebApiServer.SystemTests;

    public class OAuthTests :
        IClassFixture<WebApiServerFixture>
    {
        private readonly WebApiServerFixture server;

        public OAuthTests(WebApiServerFixture server)
        {
            this.server = server;
        }

        [Fact]
        public async Task SomeTest()
        {
            var client = server.CreateClient();

            var response = await client.GetAsync("not/an/existing/route");

            response.EnsureSuccessStatusCode();
        }
    }

**Reply by br3nt** [reply]
I also wanted to be able to do this, and have struggled to create a subclass of `WebApplicationFactory` which allows me to have fine control over configuring services and the `WebApplication`.  It seems none of the overridable methods on `WebApplicationFactory` use the same types found in `Program.cs` when using `WebApplication.CreateBuilder(args)`.

The approach found in the Microsoft docs to [Customize the client with WithWebHostBuilder](https://learn.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-8.0#customize-the-client-with-withwebhostbuilder) seems like it could be prone to error.  E.g. If I'm using a third-party library like MassTransit, in my Program.cs I have a line `services.AddMassTransit(config => { /* ... */ })`.  I don't know (or want to know) all the things that method is adding to the DI container, and I don't want to have to find the service descriptors and manually remove them either.  I'd much rather not add MassTransit in my tests in the first place, or perhaps I only want to mock the interfaces I use in my app, like `IBus`, and provide my implementation.  I could likely stuff up the running app if I don't replace hidden third-party configuration correctly.  However, I can easily ensure my application works as expected if I only provide the services used by my code under test.

All I wanted to be able to do was create an interface or abstract class that configures my web app so that in my tests I can provide alternative implementations.

I came up with this approach:

```c#
// Program.cs

var builder = WebApplication.CreateBuilder(args);
var app = builder.Build<ServicesConfiguration, WebAppConfiguration>();
app.Run();

public partial class Program {}

public static class WebApplicationBuilderExtensions
{
    public static WebApplication Build<TServicesConfiguration, TWebAppConfiguration>(this WebApplicationBuilder builder)
        where TServicesConfiguration : IServicesConfiguration
        where TWebAppConfiguration : IWebAppConfiguration
    {
        builder.Configure<TServicesConfiguration>();
        var app = builder.Build<TWebAppConfiguration>();
        return app;
    }

    public static void Configure<TServicesConfiguration>(this WebApplicationBuilder builder) where TServicesConfiguration : IServicesConfiguration
    {
        var servicesConfiguration = TServicesConfiguration.Create(builder);
        servicesConfiguration.Configure(builder.Services);
    }

    public static WebApplication Build<TWebAppConfiguration>(this WebApplicationBuilder builder) where TWebAppConfiguration : IWebAppConfiguration
    {
        var app = builder.Build();
        var webAppConfiguration = TWebAppConfiguration.Create(builder);
        webAppConfiguration.Configure(app);

        return app;
    }
}

public interface IServicesConfiguration
{
    /// <summary>Creates an instance of this subclass</summary>
    static abstract IServicesConfiguration Create(WebApplicationBuilder builder);

    public void Configure(IServiceCollection services);

    // TODO: Use Template Method pattern.  See: https://refactoring.guru/design-patterns/template-method
}

public interface IWebAppConfiguration
{
    /// <summary>Creates an instance of this subclass</summary>
    static abstract IWebAppConfiguration Create(WebApplicationBuilder builder);

    public void Configure(WebApplication app);

    // TODO: Use Template Method pattern.  See: https://refactoring.guru/design-patterns/template-method
}

public record ServicesConfiguration(IConfiguration Configuration, IWebHostEnvironment Environment) : IServicesConfiguration
{
    public static IServicesConfiguration Create(WebApplicationBuilder builder) => new ServicesConfiguration(builder.Configuration, builder.Environment);

    public void Configure(IServiceCollection services)
    {
        services.AddControllersWithViews();
    }
}

public record WebAppConfiguration(IConfiguration Configuration, IWebHostEnvironment Environment) : IWebAppConfiguration
{
    /// <inheritdoc />
    public static IWebAppConfiguration Create(WebApplicationBuilder builder) => new WebAppConfiguration(builder.Configuration, builder.Environment);

    public void Configure(WebApplication app)
    {
        if (!app.Environment.IsDevelopment())
        {
            app.UseExceptionHandler("/Home/Error");
            // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
            app.UseHsts();
        }

        app.UseHttpsRedirection();
        app.UseStaticFiles();

        app.UseRouting();

        app.UseAuthorization();

        app.MapControllerRoute(
            name: "default",
            pattern: "{controller=Home}/{action=Index}/{id?}");
    }
}
```

This code seems convoluted, but in theory, it allows me to use the [Template Method pattern](https://refactoring.guru/design-patterns/template-method).  I could modify `IServicesConfiguration` and `IWebAppConfiguration` and add methods such as `ConfigureDatabase()`, `ConfigureMassTransit()`, etc.  Then implement/override these in subclasses used in my tests.  The only problem is I couldn't figure out a way to get this to work with `WebApplicationFactory`.

Thanks to the implementation provided by @BrunoJuchli I should now be able to use this approach.

Initially, was confused by `IAsyncLifetime`.  This is an interface from XUnit.  I assume that `IClassFixture<WebApiServerFixture>` causes `InitializeAsync()` to be called before the instance of `WebApiServerFixture` is passed to the constructor of `OAuthTests`.

**Reply by br3nt** [reply]
I finally have this working.

To follow on from my previous code, I have created a new class based off the functionality of `AbstractTestServerFixture` provided by @BrunoJuchli.  My new class is a generic which requires two types that must implement `IServicesConfiguration` and `IWebAppConfiguration`.  This allows me to configure my test  server specifically for each test or set of tests by providing alternate implementations of `IServicesConfiguration` and `IWebAppConfiguration`.

Here is the code below:

```c#
public class TestWebApplication<TServicesConfiguration, TWebAppConfiguration> : IDisposable
    where TServicesConfiguration : IServicesConfiguration
    where TWebAppConfiguration : IWebAppConfiguration
{
    private CancellationTokenSource WebAppCancellationTokenSource { get; } = new();
    private WebApplication WebApplication { get; }
    private TestServer TestServer { get; }
    private List<HttpClient> HttpClients { get; } = new();

    public TestWebApplication()
    {
        var builder = WebApplication.CreateBuilder();

        builder
            .WebHost
            .UseShutdownTimeout(TimeSpan.FromSeconds(5))
            .UseTestServer();

        WebApplication = builder.Build<TServicesConfiguration, TWebAppConfiguration>();
        WebApplication.StartAsync(WebAppCancellationTokenSource.Token).GetAwaiter();

        TestServer = WebApplication.GetTestServer();
    }

    public HttpClient CreateClient() => WebApplication.GetTestClient();

    // MIT License
    // loosely based on the source code of https://www.nuget.org/packages/Microsoft.AspNetCore.Mvc.Testing
    // found at https://github.com/dotnet/aspnetcore/tree/main/src/Mvc/Mvc.Testing/src
    // (version at the time of writing: https://github.com/dotnet/aspnetcore/blob/v8.0.1/src/Mvc/Mvc.Testing/src/WebApplicationFactory.cs)
    private HttpClient CreateClientInternal(params DelegatingHandler[] handlers)
    {
        HttpClient client;
        if (handlers.Length == 0)
        {
            client = TestServer.CreateClient();
        }
        else
        {
            for (var i = handlers.Length - 1; i > 0; i--)
            {
                handlers[i - 1].InnerHandler = handlers[i];
            }

            var serverHandler = TestServer.CreateHandler();
            handlers[^1].InnerHandler = serverHandler;

            client = new HttpClient(handlers[0]);
        }

        HttpClients.Add(client);

        client.BaseAddress = TestServer.BaseAddress;

        return client;
    }

    public void Dispose()
    {
        foreach (var client in HttpClients) client.Dispose();
        WebAppCancellationTokenSource.Cancel();
        WebApplication.StopAsync(TimeSpan.FromSeconds(30)).GetAwaiter().GetResult();
        WebApplication.DisposeAsync().GetAwaiter().GetResult();
        TestServer.Dispose();
    }
}
```

For each test or set of tests, I can create a subclass of `TestWebApplication` with the specific `IServicesConfiguration` and `IWebAppConfiguration` implementations required for the test:

```c#
public class FullWebApplication : TestWebApplication<ServicesConfiguration, WebAppConfiguration> { }
```

And here is how it can be used in a unit test:

```c#
public class TestExample : IClassFixture<FullWebApplication>
{
    private FullWebApplication WebApplication { get; }

    public TestExample(FullWebApplication webApplication)
    {
        WebApplication = webApplication;
    }

    [Fact]
    public async Task HomepageReturns200ResponeWithHtmlWithUtf8Encoding()
    {
        var client = WebApplication.CreateClient();

        var response = await client.GetAsync("/");

        response.EnsureSuccessStatusCode();
        Assert.Equal("text/html; charset=utf-8", response?.Content.Headers.ContentType?.ToString());
    }
}
```

Now, when I first ran this test, it failed because the test server was responding with a 404.  This is because we are configuring the application in the context of test project rather than the web app project.  That is, the call to `services.AddControllersWithViews()` in `ServicesConfiguration` from the test project looks for controllers and views in the test project rather than the web app.  So of course no controllers or views are found.  This doesn't happen with `WebApplicationFactory` because the type provided acts as an entrypoint to the assembly which is then loaded.  With this implementation, we need to be explicit and indicate which assembly the views and controllers can be found in.  The modified code looks like this:

```c#
public record ServicesConfiguration(IConfiguration Configuration, IWebHostEnvironment Environment) : IServicesConfiguration
{
    public static IServicesConfiguration Create(WebApplicationBuilder builder) =>
        new ServicesConfiguration(builder.Configuration, builder.Environment);

    public void Configure(IServiceCollection services)
    {
        services
            .AddControllersWithViews()
            // this explicitly loads the controllers in this project to ensure they are included
            // when running integration tests
            .AddApplicationPart(GetType().Assembly);
    }
}
```

With this change to `ServicesConfiguration`, the test should now pass! 🎉

The remaining tasks for me are:
* create a subclass of `IServicesConfiguration` which configures an in memory database with a unique db name
* create and migrate the database  at startup of the test server, and drop it in `Dispose()`
* seed the database at at startup of the test server (start state for each test using this subclass)
* open a transaction before each test
* roll back the transaction after each test

[a comment by an automated account is not shown]

**Comment by Jorge0998** [comment]
Is there a way to pass additional command line arguments to the test host? Besides the ones that are being sent by default such as port, endpoint, role, etc.
